ert_data/loader.py

```diff
@@ -8,101 +8,215 @@ def data_loader_factory(observation_type):
-        return load_block_data
+        return _load_block_data
     else:
         raise TypeError("Unknown observation type: {}".format(observation_type))
 
 
-def load_general_data(facade, observation_key, case_name, include_data=True):
-    obs_vector = facade.get_observations()[observation_key]
-    data_key = obs_vector.getDataKey()
+def _extract_data(
+    facade, obs_keys, case_name, response_loader, obs_loader, expected_obs, include_data
+):
+    if isinstance(obs_keys, str):
+        obs_keys = [obs_keys]
+    data_map = {}
+    obs_types = [facade.get_impl_type_name_for_obs_key(key) for key in obs_keys]
+    data_keys = [facade.get_data_key_for_obs_key(key) for key in obs_keys]
+    assert (
+        len(set(obs_types)) == 1
+    ), f"\nExpected only {expected_obs} observation type. Found: {obs_types} for {obs_keys}"
+    assert (
+        len(set(data_keys)) == 1
+    ), f"\nExpected all obs keys ({obs_keys}) to have the same data key, found: {data_keys} "
 
-    data = pd.DataFrame()
+    if include_data:
+        # Because all observations in this loop are pointing to the same data key,
+        # we can use any of them as input to the response loader.
+        data = response_loader(facade, obs_keys[0], case_name)
+        data.columns = _create_multi_index(data.columns.to_list(), list(range(len(data.columns))))
+    else:
+        data = None
+    obs = obs_loader(facade, obs_keys, case_name)
+    if obs.empty:
+        raise NoObservationsException("No observations loaded")
+    for obs_key in obs_keys:
+        data_for_key = _get_data(data, obs[obs_key])
+        data_map[obs_key] = pd.concat([obs[obs_key], data_for_key])
 
-    for time_step in obs_vector.getStepList().asList():
-        # Fetch, then transpose the simulation data in order to make it
-        # conform with the GenObservation data structure.
-
-        # Observations and its standard deviation are a subset of the simulation data.
-        # The index_list refers to indices in the simulation data. In order to
-        # join these data in a DataFrame, pandas inserts the obs/std
-        # data into the columns representing said indices, before appending
-        # the simulation data.
-        # You then get something like:
-        #      0   1   2
-        # OBS  NaN NaN 42
-        # STD  NaN NaN 4.2
-        #   0  1.2 2.3 7.5
-        node = obs_vector.getNode(time_step)
-        index_list = [node.getIndex(nr) for nr in range(len(node))]
-
-        data = data.append(
-            pd.DataFrame([node.get_data_points()], columns=index_list, index=["OBS"])
-        ).append(pd.DataFrame([node.get_std()], columns=index_list, index=["STD"]))
-        if include_data:
-            gen_data = facade.load_gen_data(case_name, data_key, time_step).T
-            data = data.append(gen_data)
-    return data
+    return pd.concat(data_map, axis=1).astype(float)
+
+
+def _load_general_data(facade, obs_keys, case_name, include_data=True):
```

Comment:
Remind me again why we're making these private? It's already and API breach, and all we do is forward the call? 

If we really want to remove the public methods, could we for instance instead change the `data_loader_factory`?:

```
if observation_type in ["GEN_OBS", "SUMMARY_OBS", "BLOCK_OBS"]:
    return partial(_extract_data, expected_obs=observation_type)
```

Otherwise I think we shouldn't change the API.

Reply:
The reason for making them private is they were never intended to be public, the entry point was always intended to be `MeasuredData`.

We could do the second part,but would have to include the loader functions, so guess something like this?
```
    if observation_type in ["GEN_OBS", "SUMMARY_OBS", "BLOCK_OBS"]:
        if observation_type == "GEN_OBS":
            response_loader = _load_general_response
            obs_loader = _load_general_obs
        elif observation_type == "SUMMARY_OBS":
            response_loader = _load_summary_response
            obs_loader = _load_summary_obs
        elif observation_type == "BLOCK_OBS":
            response_loader = _load_block_response
            obs_loader = _load_block_obs
        return partial(_extract_data, expected_obs=observation_type, reponse_loader=response_loader, obs_loader=obs_loader)

```

Reply:
I'm not strongly opinionated here, so I'll leave it up to the author. I simply found it a bit strange to make them private and have their sole function be to pass forward to another function. I think I'd prefer use `partial` instead.

Do we otherwise want to deprecate the original functions, or silently remove them? (suppose this question goes to all changes made in this PR)

Reply:
These are the only public functions that were changed in this PR? If you disregard the last commit. But yes, can do a `partial` and just deprecate the other functions, but do you think that is necessary? Are there use cases for them? Should only be three functions?

Reply:
Yeah, it's only `load_general_data`, `load_summary_data` and `load_block_data` that are renamed/removed.

>just deprecate the other functions, but do you think that is necessary?

What are you asking is necessary? You've added the deprecation already, looks good. If `partial` complicates things I don't think it's necessary. We can add private functions in order to deprecate others, not an issue.